File: generator/spark/tpch-data-generator.scala
val args = sc.getConf.get("spark.driver.args").split("\\s+")
val compressed = args contains "compressed"
val onefile = args contains "onefile"

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.{DataFrame}

// config for compression or no compression
val newConf = if (compressed) spark.sparkContext.getConf.set("spark.sql.parquet.binaryAsString","true") else spark.sparkContext.getConf.set("spark.sql.parquet.compression.codec","none").set("spark.sql.parquet.binaryAsString","true")

// use newConf
spark.sparkContext.stop()
val spark = SparkSession.builder.config(newConf).getOrCreate()
val sc = spark.sparkContext

/// load tpch data from *.tbl files
val DATA_PATH = args(0)
val STORAGE_PATH = args(1)

/// force not nullable
def setNullableStateForAllColumns( df: DataFrame, nullable: Boolean) : DataFrame = {
  // get schema
  val schema = df.schema
  // modify [[StructField] with name `cn`
  val newSchema = StructType(schema.map {
    case StructField( c, t, _, m) ⇒ StructField( c, t, nullable = nullable, m)
  })
  // apply new schema
  df.sqlContext.createDataFrame( df.rdd, newSchema )
}

def getCustomer(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    Seq(
      StructField("c_custkey", IntegerType, false),
      StructField("c_name", StringType, false),
      StructField("c_address", StringType, false),
      StructField("c_nationkey", IntegerType, false),
      StructField("c_phone", StringType, false),
      StructField("c_acctbal", DecimalType(12, 2), false),
      StructField("c_mktsegment", StringType, false),
      StructField("c_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/customer.tbl"), false)
}

def getLineitem(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("l_orderkey", IntegerType, false),
      StructField("l_partkey", IntegerType, false),
      StructField("l_suppkey", IntegerType, false),
      StructField("l_linenumber", IntegerType, false),
      StructField("l_quantity", DecimalType(12, 2), false),
      StructField("l_extendedprice", DecimalType(12, 2), false),
      StructField("l_discount", DecimalType(12, 2), false),
      StructField("l_tax", DecimalType(12, 2), false),
      StructField("l_returnflag", StringType, false),
      StructField("l_linestatus", StringType, false),
      StructField("l_shipdate", DateType, false),
      StructField("l_commitdate", DateType, false),
      StructField("l_receiptdate", DateType, false),
      StructField("l_shipinstruct", StringType, false),
      StructField("l_shipmode", StringType, false),
      StructField("l_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/lineitem.tbl"), false)
}

def getNation(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("n_nationkey", IntegerType, false),
      StructField("n_name", StringType, false),
      StructField("n_regionkey", IntegerType, false),
      StructField("n_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/nation.tbl"), false)
}

def getOrder(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("o_orderkey", IntegerType, false),
      StructField("o_custkey", IntegerType, false),
      StructField("o_orderstatus", StringType, false),
      StructField("o_totalprice", DecimalType(12, 2), false),
      StructField("o_orderdate", DateType, false),
      StructField("o_orderpriority", StringType, false),
      StructField("o_clerk", StringType, false),
      StructField("o_shippriority", IntegerType, false),
      StructField("o_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/orders.tbl"), false)
}

def getPart(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("p_partkey", IntegerType, false),
      StructField("p_name", StringType, false),
      StructField("p_mfgr", StringType, false),
      StructField("p_brand", StringType, false),
      StructField("p_type", StringType, false),
      StructField("p_size", IntegerType, false),
      StructField("p_container", StringType, false),
      StructField("p_retailprice", DecimalType(12, 2), false),
      StructField("p_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/part.tbl"), false)
}

def getPartsupp(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("ps_partkey", IntegerType, false),
      StructField("ps_suppkey", IntegerType, false),
      StructField("ps_availqty", IntegerType, false),
      StructField("ps_supplycost", DecimalType(12, 2), false),
      StructField("ps_commment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/partsupp.tbl"), false)
}

def getRegion(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("r_regionkey", IntegerType, false),
      StructField("r_name", StringType, false),
      StructField("r_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/region.tbl"), false)
}

def getSupplier(): DataFrame = {
  setNullableStateForAllColumns(spark.read.format("csv").schema(StructType(
    List(
      StructField("s_suppkey", IntegerType, false),
      StructField("s_name", StringType, false),
      StructField("s_address", StringType, false),
      StructField("s_nationkey", IntegerType, false),
      StructField("s_phone", StringType, false),
      StructField("s_acctbal", DecimalType(12, 2), false),
      StructField("s_comment", StringType, false)
    )
  )).option("delimiter", "|").load(DATA_PATH + "/supplier.tbl"), false)
}

if (onefile) {
  getRegion.repartition(1).write.parquet(STORAGE_PATH + "/region")
  getNation.repartition(1).write.parquet(STORAGE_PATH + "/nation")
  getCustomer.repartition(1).write.parquet(STORAGE_PATH + "/customer")
  getLineitem.repartition(1).write.parquet(STORAGE_PATH + "/lineitem")
  getOrder.repartition(1).write.parquet(STORAGE_PATH + "/orders")
  getPart.repartition(1).write.parquet(STORAGE_PATH + "/part")
  getPartsupp.repartition(1).write.parquet(STORAGE_PATH + "/partsupp")
  getSupplier.repartition(1).write.parquet(STORAGE_PATH + "/supplier")
} else {
  getRegion.write.parquet(STORAGE_PATH + "/region")
  getNation.write.parquet(STORAGE_PATH + "/nation")
  getCustomer.write.parquet(STORAGE_PATH + "/customer")
  getLineitem.write.parquet(STORAGE_PATH + "/lineitem")
  getOrder.write.parquet(STORAGE_PATH + "/orders")
  getPart.write.parquet(STORAGE_PATH + "/part")
  getPartsupp.write.parquet(STORAGE_PATH + "/partsupp")
  getSupplier.write.parquet(STORAGE_PATH + "/supplier")
}

File: generator/duckdb/tpch-data-generator.py
#!/usr/bin/python3
import duckdb
import timeit
import sys

if len(sys.argv) != 3:
    print("Please pass the source path of the TPC-H CSV files and the destination path where the Parquet files should be stored")
    sys.exit()

source_path = sys.argv[1]
destination_path = sys.argv[2]

con = duckdb.connect(database=':memory:', read_only=False)

initialisationQuery = ("""create table part (
   p_partkey integer not null,
   p_name varchar(55) not null,
   p_mfgr char(25) not null,
   p_brand char(10) not null,
   p_type varchar(25) not null,
   p_size integer not null,
   p_container char(10) not null,
   p_retailprice decimal(12,2) not null,
   p_comment varchar(23) not null,
   primary key (p_partkey)
);

create table supplier (
   s_suppkey integer not null,
   s_name char(25) not null,
   s_address varchar(40) not null,
   s_nationkey integer not null,
   s_phone char(15) not null,
   s_acctbal decimal(12,2) not null,
   s_comment varchar(101) not null,
   primary key (s_suppkey)
);

create table partsupp (
   ps_partkey integer not null,
   ps_suppkey integer not null,
   ps_availqty integer not null,
   ps_supplycost decimal(12,2) not null,
   ps_comment varchar(199) not null,
   primary key (ps_partkey,ps_suppkey)
);

create table customer (
   c_custkey integer not null,
   c_name varchar(25) not null,
   c_address varchar(40) not null,
   c_nationkey integer not null,
   c_phone char(15) not null,
   c_acctbal decimal(12,2) not null,
   c_mktsegment char(10) not null,
   c_comment varchar(117) not null,
   primary key (c_custkey)
);

create table orders (
   o_orderkey integer not null,
   o_custkey integer not null,
   o_orderstatus char(1) not null,
   o_totalprice decimal(12,2) not null,
   o_orderdate date not null,
   o_orderpriority char(15) not null,
   o_clerk char(15) not null,
   o_shippriority integer not null,
   o_comment varchar(79) not null,
   primary key (o_orderkey)
);

create table lineitem (
   l_orderkey integer not null,
   l_partkey integer not null,
   l_suppkey integer not null,
   l_linenumber integer not null,
   l_quantity decimal(12,2) not null,
   l_extendedprice decimal(12,2) not null,
   l_discount decimal(12,2) not null,
   l_tax decimal(12,2) not null,
   l_returnflag char(1) not null,
   l_linestatus char(1) not null,
   l_shipdate date not null,
   l_commitdate date not null,
   l_receiptdate date not null,
   l_shipinstruct char(25) not null,
   l_shipmode char(10) not null,
   l_comment varchar(44) not null,
   primary key (l_orderkey,l_linenumber)
);

create table nation (
   n_nationkey integer not null,
   n_name char(25) not null,
   n_regionkey integer not null,
   n_comment varchar(152) not null,
   primary key (n_nationkey)
);

create table region (
   r_regionkey integer not null,
   r_name char(25) not null,
   r_comment varchar(152) not null,
   primary key (r_regionkey)
);

copy part from '""" + source_path + """/part.tbl' delimiter '|';
COPY (SELECT * FROM part) TO '""" + destination_path + """/part.parquet' (FORMAT 'parquet');

copy supplier from '""" + source_path + """/supplier.tbl' delimiter '|';
COPY (SELECT * FROM supplier) TO '""" + destination_path + """/supplier.parquet' (FORMAT 'parquet');

copy partsupp from '""" + source_path + """/partsupp.tbl' delimiter '|';
COPY (SELECT * FROM partsupp) TO '""" + destination_path + """/partsupp.parquet' (FORMAT 'parquet');

copy customer from '""" + source_path + """/customer.tbl' delimiter '|';
COPY (SELECT * FROM customer) TO '""" + destination_path + """/customer.parquet' (FORMAT 'parquet');

copy orders from '""" + source_path + """/orders.tbl' delimiter '|';
COPY (SELECT * FROM orders) TO '""" + destination_path + """/orders.parquet' (FORMAT 'parquet');

copy lineitem from '""" + source_path + """/lineitem.tbl' delimiter '|';
COPY (SELECT * FROM lineitem) TO '""" + destination_path + """/lineitem.parquet' (FORMAT 'parquet');

copy nation from '""" + source_path + """/nation.tbl' delimiter '|';
COPY (SELECT * FROM nation) TO '""" + destination_path + """/nation.parquet' (FORMAT 'parquet');

copy region from '""" + source_path + """/region.tbl' delimiter '|';
COPY (SELECT * FROM region) TO '""" + destination_path + """/region.parquet' (FORMAT 'parquet');

""")

con.execute(query=initialisationQuery)